contracts/FastBreak.cdc

```diff
@@ -0,0 +1,682 @@
+/*
+    Fast Break Game Contract
+    Author: Jeremy Ahrens jer.ahrens@dapperlabs.com
+*/
+
+import NonFungibleToken from 0xNFTADDRESS
+import TopShot from 0xTOPSHOTADDRESS
+
+/// Game & Oracle Contract for Fast Break
+///
+pub contract FastBreak: NonFungibleToken {
+
+    /// Contract events
+    ///
+    pub event ContractInitialized()
+    pub event Withdraw(id: UInt64, from: Address?)
+    pub event Deposit(id: UInt64, to: Address?)
+    pub event FastBreakRunCreated(
+        id: String,
+        name: String,
+        runStart: UInt64,
+        runEnd: UInt64,
+        fatigueModeOn: Bool
+    )
+    pub event FastBreakRunStatusChange(id: String, newStatus: String)
+    pub event FastBreakGameCreated(
+        id: String,
+        name: String,
+        fastBreakRunID: String,
+        isPublic: Bool,
+        submissionDeadline: UInt64,
+        numPlayers: UInt64
+    )
+    pub event FastBreakGameStatusChange(id: String, newStatus: String)
+    pub event FastBreakNFTBurned(id: UInt64, serialNumber: UInt64)
+    pub event FastBreakNFTMinted(
+        id: UInt64,
+        fastBreakGameID: String,
+        serialNumber: UInt64,
+        mintingDate: UInt64,
+        topShots: [UInt64],
+        mintedTo: Address
+    )
+    pub event FastBreakSubmissionSent(
+        wallet: Address,
+        submittedAt: UInt64,
+        fastBreakGameID: String,
+        topShots: [UInt64]
+    )
+    pub event FastBreakGameWinner(
+        wallet: Address?,
+        submittedAt: UInt64,
+        fastBreakGameID: String,
+        topShots: [UInt64]
+    )
+    pub event FastBreakGameStatAdded(
+        fastBreakGameID: String,
+        name: String,
+        type: String,
+        valueNeeded: UInt64
+    )
+
+    /// Named Paths
+    ///
+    pub let CollectionStoragePath:  StoragePath
+    pub let CollectionPublicPath:   PublicPath
+    pub let OracleStoragePath:       StoragePath
+    pub let OraclePrivatePath:      PrivatePath
+
+    /// Contract variables
+    ///
+    pub var totalSupply:        UInt64
+
+    /// Metadata Dictionaries
+    ///
+    access(self) let fastBreakRunByID:        {String: FastBreakRun}
+    access(self) let fastBreakGameByID:           {String: FastBreakGame}
```

Comment:
No idea about the business logic, so may not apply/be possible: Maybe consider refactoring the architecture of how/where games are stored. Currently, the contract itself stores all games, which might become a scaling bottleneck. Is this really necessary?

Maybe try and see if account-based storage architecture could work here. For example, for a game with a "tournament"/"round", allow it to be created and stored in any account.

Reply:
agreed with this. It might be useful for users to store the games they participate in in their own account or something like that

Reply:
Yes, I agree and that's what I was going to do initially. It's more ideal for sure.

The problem that I could not figure out how to solve is that the game score for each user will be updated 100s of times by the oracle as the NBA games are going on. I can write to the token for custodial accounts I have the key for but don't think I can write to a non-custodial account token. Non-custodial accounts would not be able to get their token scored, afaik.

I don't write a ton of cadence so perhaps I am missing something, but wanted to explain why I took the direction of game object storage in the contract as opposed to wallets.

Reply:
Also not knowing the inner-workings of Flow: does it change anything if the games and runs are resources owned by the admin wallet (instead of a struct)? Then when a tx needs to fetch details/scores about the game we would do a borrow operation on a specific game.

Reply:
yeah, probably does make sense to just store them centrally.

Reply:
Resolving this convo as it appears a central ledger is operative here for the time being.